### PR TITLE
Define a reason why the fifocache has evicted an item

### PR DESCRIPTION
**Pull request by dannykopping**
Signed-off-by: Danny Kopping <danny.kopping@grafana.com>

**What this PR does / why we need it**:
This PR adds an eviction reason for the fifocache, which will help an operator know if an entry was evicted because the TTL expired vs the cache is full.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:
We use a fifocache when running boltdb-shipper for storing ingesters' index caches locally. We do this because we cannot share the cache between ingesters, to prevent missing out chunks that may not exist in other ingesters.
https://github.com/grafana/loki/blob/main/pkg/loki/modules.go#L414-L425

The cache is configured with a fixed size of 200MB, and is currently not configurable; we could consider making this configurable however even in our biggest clusters we're only seeing ~11MB of use currently.

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: pkg/storage/chunk/cache/fifo_cache.go [hunk 1]
```diff
@@ -86,14 +86,21 @@ type FifoCache struct {
...
 	totalMisses     prometheus.Counter
 	staleGets       prometheus.Counter
 	memoryBytes     prometheus.Gauge
 }
 
+const (
+	expiredReason string = "expired" //nolint:staticcheck
+	fullReason           = "full"
+	tooBigReason         = "too big"
+	stoppedReason        = "stopped"
```
Comment (kavirajk) [inline pkg/storage/chunk/cache/fifo_cache.go:101]:
```suggestion
	expiredReason string = "object-expired" //nolint:staticcheck
	fullReason           = "cache-full"
	tooBigReason         = "object-too-big"
	stoppedReason        = "cache-stopped"
```

**Review by kavirajk — COMMENTED** [review]
LGTM :+1: 

Minor suggestions on `reason` constants.

File: pkg/storage/chunk/cache/fifo_cache.go (hunk 1 above)
Comment (dannykopping) [reply pkg/storage/chunk/cache/fifo_cache.go:101]: Thanks for the review :+1: 

I don't think these suggestions make things much more clear TBH.
I agree on this one though `tooBigReason = "object-too-big"`.
Can you expand on why you think the others make things more clear pls?

File: pkg/storage/chunk/cache/fifo_cache.go [hunk 2]
```diff
@@ -86,14 +86,21 @@ type FifoCache struct {
...
 	totalMisses     prometheus.Counter
 	staleGets       prometheus.Counter
 	memoryBytes     prometheus.Gauge
 }
 
+const (
+	expiredReason string = "expired" //nolint:staticcheck
+	fullReason           = "full"
+	tooBigReason         = "object too big"
+	stoppedReason        = "stopped"
```
Comment (sandeepsukhani) [inline pkg/storage/chunk/cache/fifo_cache.go:101]: `stopped` doesn't give the operator much context, especially the ones who do not go and check the code. I think we should not increment the metric when we are evicting items while stopping. It doesn't make a lot of sense to me.

**Review by sandeepsukhani — APPROVED** [review]
LGTM with a minor non-blocking suggestion

File: pkg/storage/chunk/cache/fifo_cache.go (hunk 2 above)
Comment (dannykopping) [reply pkg/storage/chunk/cache/fifo_cache.go:101]: I thought the same actually, cool. Will remove :+1:

File: pkg/storage/chunk/cache/fifo_cache.go (hunk 1 above)
Comment (kavirajk) [reply pkg/storage/chunk/cache/fifo_cache.go]: if `stopped` is removed. Then other looks ok :)

**Review by kavirajk — APPROVED** [review]

File: pkg/storage/chunk/cache/fifo_cache.go (hunk 1 above)
Comment (dannykopping) [reply pkg/storage/chunk/cache/fifo_cache.go]: Discussed offline, Kavi's satisfied with my change

**Pull request merged**